### ISSUE TITLE
docker daemon locks files in /dev

### ISSUE THREAD

**Issue by nolange** [moby/moby#31182]
**Description**

A running docker daemon will lock files in /dev, other tools that want to temporary take such looks will block forever or fail.

**Steps to reproduce the issue:**

```
# run with
#
# rm -rf subdir
# make -O >/dev/null
#
# add -j2, then you can' t even terminate it

FILE := a b

all: $(addprefix subdir/,$(FILE))

$(addprefix subdir/,$(FILE)):
	echo "$@"
	mkdir -p $(dir $@); touch $@
```
1. Have the dockerd daemon running
2. Run the above Makefile with `make -O >/dev/null`

**Describe the results you received:**

make finishing

**Describe the results you expected:**

make is blocked waiting for a lock on /dev/null indefinitely. Stopping the dockerd service will allow it to continue.

**Additional information you deem important (e.g. issue happens only occasionally):**

Issue first appeared on January 19th, using the docker reporsitory `deb https://apt.dockerproject.org/repo debian-stretch main`
[Bug report for debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=855336)

**Output of `docker version`:**

```
Client:
 Version:      1.13.1
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   092cba372
 Built:        Wed Feb  8 06:44:30 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.1
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   092cba372
 Built:        Wed Feb  8 06:44:30 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 9
Server Version: 1.13.1
Storage Driver: devicemapper
 Pool Name: docker-8:2-12328161-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 6.814 GB
 Data Space Total: 107.4 GB
 Data Space Available: 100.6 GB
 Metadata Space Used: 5.927 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.142 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.137 (2016-11-30)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-1-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 15.62 GiB
Name: desbuild
ID: 24GS:G7V4:MEKN:PBVA:MED3:JEMB:E5KG:5JHI:2SH6:7IGU:DOYL:Y4KH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by justincormack** [comment]
cc @ijc25

**Comment by nolange** [comment]
bug is still unresolved, killed another builder.

lslocks:
dockerd           730 OFDLCK      READ  0     0   0 /dev...

**Comment by mailinglists35** [comment]
this bug appears to make random victims. sudo, this time: https://old.reddit.com/r/linuxadmin/comments/9ag0v0/sudo_hangs_forever_after_distro_upgrade/

**Comment by moritzboth** [comment]
[amanda](http://www.amanda.org/) won't run for me because it wants to flock() /dev/null for a nanosecond. Looking into amanda's source code, this is not as stupid as it sounds: It uses the lock on stdout to *portably* sync writes to a shared pipe. Other programs may use this tweak too - I don't know that.

So it would be really nice if dockerd refrained from locking something in /dev permanantly without using it

**Comment by joakim-tjernlund** [comment]
Got hit by this bug too, see https://bugs.gentoo.org/748984

**Comment by thaJeztah** [comment]
@kolyshkin @cpuguy83 @tonistiigi ptal

**Comment by joakim-tjernlund** [comment]
I don't know GO but could this be the cause:
```
func init() {
	// use open file descriptor locks if the system supports it
	getlk := syscall.Flock_t{Type: syscall.F_RDLCK}
	if err := syscall.FcntlFlock(0, F_OFD_GETLK, &getlk); err == nil {
		linuxTryLockFile = ofdTryLockFile
		linuxLockFile = ofdLockFile
	}
}
```
The test for ODF locks leaves behind a lock?

**Comment by joakim-tjernlund** [comment]
This dummy code make the lock go away:
```
--- docker-ce/components/engine/vendor/github.com/coreos/etcd/pkg/fileutil/lock_linux.go	2020-10-31 16:33:59.855609583 +0100
+++ docker-ce/components/engine/vendor/github.com/coreos/etcd/pkg/fileutil/lock_linux.go	2020-10-31 16:34:13.607556963 +0100
@@ -48,11 +48,11 @@
 
 func init() {
 	// use open file descriptor locks if the system supports it
-	getlk := syscall.Flock_t{Type: syscall.F_RDLCK}
-	if err := syscall.FcntlFlock(0, F_OFD_GETLK, &getlk); err == nil {
+	//getlk := syscall.Flock_t{Type: syscall.F_RDLCK}
+	//if err := syscall.FcntlFlock(0, F_OFD_GETLK, &getlk); err == nil {
 		linuxTryLockFile = ofdTryLockFile
 		linuxLockFile = ofdLockFile
-	}
+	//}
 }
 
 func TryLockFile(path string, flag int, perm os.FileMode) (*LockedFile, error) {
```
Not sure if F_OFD_GETLK should create a lock or just report ? Maybe docker should fins a better way of testing for OFD locks?

**Comment by joakim-tjernlund** [comment]
Perhaps it will be enough to just replace 
  getlk := syscall.Flock_t{Type: syscall.F_RDLCK}
with
  getlk := syscall.Flock_t{Type: syscall.F_UNLCK}
?

**Comment by joakim-tjernlund** [comment]
Here is a type I think:
```
const (
	F_OFD_GETLK  = 37
	F_OFD_SETLK  = 37
	F_OFD_SETLKW = 38
)
```
F_OFD_GETLK should be 36

**Comment by moritzboth** [comment]
@joakim-tjernlund thank you so much spending your Sunday digging so deep

if this is the fix it made my day

**Comment by tonistiigi** [comment]
@joakim-tjernlund nice find. could you open up a pr in https://github.com/etcd-io/etcd to fix this?

**Comment by joakim-tjernlund** [comment]
> @joakim-tjernlund thank you so much spending your Sunday digging so deep
> 
> if this is the fix it made my day

Thanks, I figure I could not wait another 3 years for a solution :) 
The extra lock went away when I changed F_OFD_GETLK to 36

> @joakim-tjernlund nice find. could you open up a pr in https://github.com/etcd-io/etcd to fix this?

I think the devs can change a constant without me going through the hassle to do a PR as well.

**Comment by moritzboth** [comment]
I took the liberty to create the pull request. 
https://github.com/etcd-io/etcd/pull/12440 
I hope that is okay and I'm glad to be able to contribute with some simple work.

**Comment by joakim-tjernlund** [comment]
> I took the liberty to create the pull request.
> [etcd-io/etcd#12440](https://github.com/etcd-io/etcd/pull/12440)
> I hope that is okay and I'm glad to be able to contribute with some simple work.

Sure :)

**Comment by kolyshkin** [comment]
Alas, etcd did not backport my (or @moritzboth) PR to the released version :( so I opened a few backport PRs.

To fix this in Docker, we need to either wait till etcd will merge a backport PR and release a version, or use a temporary fork like I do in https://github.com/moby/moby/pull/41791.

**Comment by williamh** [comment]
Hi all,

I am one of the maintainers of the docker packages on Gentoo linux. We have been hitting this issue for some time (the downstream bug report is here [1]).

I have been asked to look into backporting some kind of fix for this to docker 20.10.2, but I want to find out your thoughts on me attempting to do this. Is changing the constant in the vendored etcd enough, or does it require updating the vendored etcd as in the pull request above?

If changing the constant isn't enough, can you give me some idea of when you will look at bumping etcd and doing a new release of the docker engine with the new etcd?

Thanks for your time and your thoughts on this.

[1] https://bugs.gentoo.org/748984

**Comment by thaJeztah** [comment]
We are waiting for  https://github.com/etcd-io/etcd/pull/12552 to be accepted upstream, and to be in a release, after that we can update the vendoring (there's a draft pull request here; https://github.com/moby/moby/pull/41791)
